File: ga_py/benchmark/gpr_overhead_bench.cpp
// C++ baseline for the Python-dispatch overhead benchmark.
//
// Measures the wall-clock cost of N = 50 000 geometric products
// `vec3d * vec3d -> mvec3d_e` in a tight loop, with no Python in sight.
// The matching Python script (gpr_overhead_bench.py) does the same N
// products via the ga_py bindings, in pure Python, and via numpy — the
// ratio of times answers "what does Python cost me for this calculation?"
//
// Build mode matters: Debug (`-g`) vs Release (`-O3 -DNDEBUG`) flow
// through CMAKE_BUILD_TYPE; pair Debug-vs-Debug and Release-vs-Release for
// a fair comparison with the Python side. The Python extension `_ga_py`
// is built with the same flags out of the same CMake configure.
//
// Anti-DCE: a volatile sink accumulates one component of every product so
// the compiler can't eliminate the loop body under -O3.
//
// Run: build/ga_py/benchmark/gpr_overhead_bench [N] [repeats]
//   defaults: N = 50000, repeats = 7 (best-of-N is reported)

#include "ga/ga_ega.hpp"

#include <fmt/format.h>

#include <chrono>
#include <cstdlib>
#include <vector>

using namespace hd::ga;       // type aliases (vec3d, mvec3d_e, …)
using namespace hd::ga::ega;  // EGA-specific operators (operator*, gpr, …)

namespace {

// Deterministic data, same shape as the Python side. We don't need actual
// randomness — the timing of `gpr` is data-independent (no branches), and
// reproducibility across runs matters more.
std::vector<vec3d> make_vecs(std::size_t N, double offset)
{
    std::vector<vec3d> out;
    out.reserve(N);
    for (std::size_t i = 0; i < N; ++i) {
        double const x = static_cast<double>(i) * 0.001 + offset;
        out.emplace_back(x, x + 0.5, x + 1.0);
    }
    return out;
}

// One timed pass over the whole input; returns elapsed seconds.
double one_pass(std::vector<vec3d> const& as,
                std::vector<vec3d> const& bs,
                volatile double& sink)
{
    auto const t0 = std::chrono::steady_clock::now();
    double local_sink = 0.0;
    for (std::size_t i = 0; i < as.size(); ++i) {
        // operator* on (vec3d, vec3d) returns mvec3d_e: scalar + bivector.
        // Field layout: c0 = scalar; c1, c2, c3 = bivector components.
        mvec3d_e const c = as[i] * bs[i];
        local_sink += c.c0 + c.c1 + c.c2 + c.c3;
    }
    auto const t1 = std::chrono::steady_clock::now();
    sink = local_sink;  // observable side-effect → loop body cannot be DCE'd
    return std::chrono::duration<double>(t1 - t0).count();
}

}  // namespace


int main(int argc, char** argv)
{
    std::size_t N = (argc > 1) ? std::strtoull(argv[1], nullptr, 10) : 50000;
    int repeats   = (argc > 2) ? std::atoi(argv[2]) : 7;

    auto const as = make_vecs(N, 0.0);
    auto const bs = make_vecs(N, 100.0);

    volatile double sink = 0.0;
    double best = 1e9;
    for (int r = 0; r < repeats; ++r) {
        double const t = one_pass(as, bs, sink);
        if (t < best) best = t;
    }

    double const ns_per_call = best * 1e9 / static_cast<double>(N);

    fmt::print("# C++ gpr overhead — N={} elements, best of {}\n", N, repeats);
    fmt::print("# vec3d * vec3d -> mvec3d_e  (geometric product)\n");
    fmt::print("\n");
    fmt::print("  C++ tight loop:    {:7.2f} ms total   {:6.1f} ns/call\n",
               best * 1e3, ns_per_call);
    fmt::print("  (sink={:.6e} — anti-DCE, ignore the value)\n",
               static_cast<double>(sink));
    return 0;
}

File: ga_py/benchmark/gpr_overhead_bench.py
"""Quantify the Python-dispatch tax for a tight loop of geometric products.

Companion to gpr_overhead_bench.cpp. Both run N = 50 000 geometric products
`vec3d * vec3d -> mvec3d_e` with deterministic data, so the C++/Python ratio
isolates the cost of "calling out to C++ N times from Python" for a calc
that's small per call (vec3d gpr is ~9 multiplies + 6 adds + 4-double result).

Three Python paths are timed:

  1. ga_py bindings        — one C++ call per element via `a * b`. The dispatch
                             tax is the fee you pay per element.
  2. pure Python (floats)  — hand-coded gpr formula on float tuples. Shows what
                             "no C++ at all" costs in CPython.
  3. numpy vectorised      — same calc, but as broadcast operations on a
                             pre-flattened (N, 3) array. The "right way" to do
                             this in Python once you commit to numpy.

Run from the project root, with the wrapper venv active and `_ga_py` built:

    PYTHONPATH=build/ga_py:ga_py/python \\
        ga_py/.venv/bin/python ga_py/benchmark/gpr_overhead_bench.py

For a fair comparison, run the C++ side in the same build mode:

    build/ga_py/benchmark/gpr_overhead_bench

Captured runs live in ga_py/benchmark/gpr_overhead_results_{debug,release}.txt.
"""
from __future__ import annotations

import sys
import timeit

import numpy as np

import ga_py


N = 50_000
REPEAT = 7


def fmt_ms(t: float) -> str:
    return f"{t * 1e3:7.2f} ms"


def fmt_ns(t: float, n: int) -> str:
    return f"{t * 1e9 / n:6.1f} ns/call"


def make_vecs(offset: float) -> list:
    """Match the C++ side's data: linearly increasing x with given offset.

    Vec3d(i*0.001 + offset, i*0.001 + offset + 0.5, i*0.001 + offset + 1.0)
    for i in [0, N).
    """
    return [ga_py.ega.vec3d(i * 0.001 + offset,
                            i * 0.001 + offset + 0.5,
                            i * 0.001 + offset + 1.0)
            for i in range(N)]


def make_float_tuples(offset: float) -> list:
    return [(i * 0.001 + offset,
             i * 0.001 + offset + 0.5,
             i * 0.001 + offset + 1.0)
            for i in range(N)]


def make_np_array(offset: float) -> np.ndarray:
    i = np.arange(N, dtype=np.float64)
    x = i * 0.001 + offset
    return np.stack([x, x + 0.5, x + 1.0], axis=1)  # (N, 3)


# --- Path 1: ga_py bindings ---------------------------------------------------

def gpr_via_bindings(as_, bs):
    """One C++ call per element via the bound `__mul__`."""
    return [a * b for a, b in zip(as_, bs)]


# --- Path 2: pure Python -----------------------------------------------------

def gpr_pure_python(as_, bs):
    """Hand-coded geometric product on float tuples, no C++ at all.

    For vec3d a, b in EGA3D, a * b = (a · b) + (a ∧ b), giving:
      scalar:        ax*bx + ay*by + az*bz
      bivec yz part: ay*bz - az*by
      bivec zx part: az*bx - ax*bz
      bivec xy part: ax*by - ay*bx
    Returned as a 4-tuple (s, byz, bzx, bxy) — same shape as mvec3d_e's
    field layout.
    """
    out = []
    append = out.append
    for (ax, ay, az), (bx, by, bz) in zip(as_, bs):
        s = ax * bx + ay * by + az * bz
        byz = ay * bz - az * by
        bzx = az * bx - ax * bz
        bxy = ax * by - ay * bx
        append((s, byz, bzx, bxy))
    return out


# --- Path 3: numpy vectorised ------------------------------------------------

def gpr_numpy(a_arr: np.ndarray, b_arr: np.ndarray) -> np.ndarray:
    """Vectorised gpr: (N, 3) × (N, 3) -> (N, 4) with [s, byz, bzx, bxy]."""
    ax, ay, az = a_arr[:, 0], a_arr[:, 1], a_arr[:, 2]
    bx, by, bz = b_arr[:, 0], b_arr[:, 1], b_arr[:, 2]
    return np.stack([
        ax * bx + ay * by + az * bz,
        ay * bz - az * by,
        az * bx - ax * bz,
        ax * by - ay * bx,
    ], axis=1)


def bench(label: str, fn) -> float:
    fn()  # warmup
    t = min(timeit.repeat(fn, number=1, repeat=REPEAT))
    print(f"  {label:24s} {fmt_ms(t)}   {fmt_ns(t, N)}")
    return t


def main() -> None:
    print(f"# Python gpr overhead — N={N} elements, best of {REPEAT}")
    print(f"# Python {sys.version.split()[0]}, NumPy {np.__version__}")
    print(f"# vec3d * vec3d -> mvec3d_e  (geometric product)")
    print()

    as_bound = make_vecs(0.0)
    bs_bound = make_vecs(100.0)
    as_tup = make_float_tuples(0.0)
    bs_tup = make_float_tuples(100.0)
    a_np = make_np_array(0.0)
    b_np = make_np_array(100.0)

    t_bind = bench("ga_py bindings",
                   lambda: gpr_via_bindings(as_bound, bs_bound))
    t_pure = bench("pure Python",
                   lambda: gpr_pure_python(as_tup, bs_tup))
    t_np = bench("numpy vectorised",
                 lambda: gpr_numpy(a_np, b_np))

    print()
    print("  ratios (lower is better):")
    print(f"    bindings / numpy        = {t_bind / t_np:6.1f}×")
    print(f"    pure python / numpy     = {t_pure / t_np:6.1f}×")
    print(f"    bindings / pure python  = {t_bind / t_pure:6.2f}×")


if __name__ == "__main__":
    main()
